torch/_inductor/fx_passes/b2b_gemm.py

```diff
@@ -97,175 +188,406 @@ def b2b_gemm_grid(M, P, meta):
+    # the inner MM
+    inner_mm = match.nodes[-1]
+
+    # find the (candidate) outer MM, which will be re-checked below to ensure every path reaches it
+    # In a real (A @ f(B @ C)), every path starting from (B @ C) must reach (A @ _).
+    outer_mm = None
+    node = inner_mm
+    while len(node.users) > 0:
+        node = next(iter(node.users))
+        if is_mm(node):
+            outer_mm = node
+            break
+        elif is_pointwise_node(node):
+            continue
+        else:
+            break
+    if not outer_mm:
+        return
+
+    # find the unique input node for outer_mm representing f(B @ C) in (A @ f(B @ C))
+    # we call it the "f_node"
+    # when the pattern is simply (A @ (B @ C)), f_node is just inner_mm
+    f_node = inner_mm
+    while next(iter(f_node.users)) is not outer_mm:
+        f_node = next(iter(f_node.users))
+
+    def all_reach_via_pointwise_with_no_other_inputs(
+        src: torch.fx.Node,
+        dst: torch.fx.Node,
+    ) -> bool:
+        """
+        check whether every user path from src reaches dst via pointwise nodes,
+        with no other input nodes for the intermediates and dst
+        """
+        if src == dst:  # trivial subgraph
+            return True
+
+        reachable: Dict[torch.fx.Node, bool] = {}
+        input_counter: Dict[torch.fx.Node, int] = {}
+
+        def reach(node: torch.fx.Node) -> bool:
+            if node in reachable:
+                return reachable[node]
+            if node is dst:
+                reachable[node] = True
+                return True
+            # for nodes other than dst, bookkeep their users' input counts
+            for user in node.users.keys():
+                if user not in input_counter:
+                    input_counter[user] = len(user.all_input_nodes)
+                input_counter[user] -= 1
+            # continue checking reachability
+            if (node is src) or is_pointwise_node(node):
+                ret = True
+                for user in node.users.keys():
+                    if not reach(user):
+                        ret = False
+                        break
+            else:
+                ret = False
+            reachable[node] = ret
+            return ret
+
+        return reach(src) and all(count == 0 for count in input_counter.values())
+
+    # check inner_mm reaches f_node on every user path via pointwise nodes with no outside input_nodes
+    if not all_reach_via_pointwise_with_no_other_inputs(inner_mm, f_node):
+        return
+
+    # check inner_mm's inputs and f_node's outputs
+    if not (len(inner_mm.all_input_nodes) == 2 and len(f_node.users) == 1):
+        return
+
+    # at this point, the nodes between inner_mm and f_node (both included)
+    # are all used internally inside (A @ subgraph(B @ C))
+    # i.e. they neither have other users nor have other inputs
+
+    # find nodes in the subgraph (including inner_mm and f_node, excluding outer_mm)
+    subgraph_node_set = set()
```

Comment:
Why don't we just reuse reachable, or input_counter ? we have three separate searches in this function, it feels like excessive complexity

Reply:
It's just more to reason about as both a reviewer and maintainer

Reply:
Removed this search by reusing `reachable`